tortoisebeacon/handlers.go

```diff
@@ -0,0 +1,213 @@
+package tortoisebeacon
+
+import (
+	"fmt"
+
+	"github.com/spacemeshos/go-spacemesh/common/types"
+	"github.com/spacemeshos/go-spacemesh/log"
+	"github.com/spacemeshos/go-spacemesh/p2p/p2pcrypto"
+	"github.com/spacemeshos/go-spacemesh/p2p/service"
+)
+
+// TBProposalProtocol is Tortoise Beacon proposal Gossip protocol name.
+const TBProposalProtocol = "TBProposalGossip"
+
+// TBVotingProtocol is Tortoise Beacon voting Gossip protocol name.
+const TBVotingProtocol = "TBVotingGossip"
+
+// HandleSerializedProposalMessage defines method to handle Tortoise Beacon proposal Messages from gossip.
+func (tb *TortoiseBeacon) HandleSerializedProposalMessage(data service.GossipMessage, sync service.Fetcher) {
+	tb.Log.With().Info("New proposal message",
+		log.String("from", data.Sender().String()))
+
+	var m ProposalMessage
+	if err := types.BytesToInterface(data.Bytes(), &m); err != nil {
+		tb.Log.With().Error("Received invalid proposal message",
+			log.String("message", string(data.Bytes())),
+			log.Err(err))
+
+		return
+	}
+
+	if err := tb.handleProposalMessage(m); err != nil {
+		tb.Log.With().Error("Failed to handle proposal message",
+			log.String("message", m.String()),
+			log.Err(err))
+
+		return
+	}
+
+	data.ReportValidation(TBProposalProtocol)
+}
+
+func (tb *TortoiseBeacon) handleProposalMessage(m ProposalMessage) error {
+	epoch := m.Epoch()
+
+	mt := tb.classifyMessage(m, epoch)
+
+	switch mt {
+	case TimelyMessage:
+		tb.Log.With().Info("Received timely ProposalMessage",
+			log.Uint64("epoch_id", uint64(m.Epoch())),
+			log.String("message", m.String()))
+
+		tb.timelyProposalsMu.Lock()
+
+		if _, ok := tb.timelyProposals[epoch]; !ok {
+			tb.timelyProposals[epoch] = make(map[types.Hash32]struct{})
+		}
+
+		tb.timelyProposals[epoch][m.Proposals().Hash()] = struct{}{}
+
+		tb.timelyProposalsMu.Unlock()
+
+		return nil
+
+	case DelayedMessage:
+		tb.Log.With().Info("Received delayed ProposalMessage",
+			log.Uint64("epoch_id", uint64(m.Epoch())),
+			log.String("message", m.String()))
+
+		tb.delayedProposalsMu.Lock()
+
+		if _, ok := tb.delayedProposals[epoch]; !ok {
+			tb.delayedProposals[epoch] = make(map[types.Hash32]struct{})
+		}
+
+		tb.delayedProposals[epoch][m.Proposals().Hash()] = struct{}{}
+
+		tb.delayedProposalsMu.Unlock()
+
+		return nil
+
+	case LateMessage:
+		tb.Log.With().Info("Received late ProposalMessage",
```

Comment:
I'd change to Warning - this is something we would potentially look for when debugging nodes